keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,750 @@
+	Size         uint64          `json:"size"`
+}
+
+// ContainerMemoryAssignments stores memory assignments of containers
+type ContainerMemoryAssignments map[string]map[string][]Block
+
+// Reader interface used to read current memory/pod assignment state
+type Reader interface {
+	// GetMachineState returns Memory Map stored in the State
+	GetMachineState() NodeMap
+	// GetMemoryBlocks returns memory assignments of a container
+	GetMemoryBlocks(podUID string, containerName string) []Block
+	// GetMemoryAssignments returns ContainerMemoryAssignments
+	GetMemoryAssignments() ContainerMemoryAssignments
+}
+
+type writer interface {
+	// SetMachineState stores NodeMap in State
+	SetMachineState(memoryMap NodeMap)
+	// SetMemoryBlocks stores memory assignments of a container
+	SetMemoryBlocks(podUID string, containerName string, blocks []Block)
+	// SetMemoryAssignments sets ContainerMemoryAssignments by using the passed parameter
+	SetMemoryAssignments(assignments ContainerMemoryAssignments)
+	// Delete deletes corresponding Blocks from ContainerMemoryAssignments
+	Delete(podUID string, containerName string)
+	// ClearState clears machineState and ContainerMemoryAssignments
+	ClearState()
+}
+
+// State interface provides methods for tracking and setting memory/pod assignment
+type State interface {
+	Reader
+	writer
+}
+
+### Changes of existing components
+
+#### Container Manager changes
+
+Container Manager will create Memory Manager and register to Topology Manager as hint provider.
+
+#### Topology Manager changes
+
+Topology Manager will call out Memory Manager to gather topology hint and allocate memory resources during its admit sequence.
+
+#### Internal Container Lifecycle changes
+
+InternalContainerLifecycle will call out Memory Manager on container lifecycle events(Add/Stop) to allocate and reclaim memory resources.
+
+### Feature Gate and Kubelet Flags
+
+A new feature gate will be added to enable the Memory Manager feature. This feature gate will be disabled by default in the early releases.
+
+#### Feature Gate  
+`--feature-gate=MemoryManager=true`
+
+#### Memory Manager Policy Flag  
+`--memory-manager-policy` - to configure Memory Manager behaviour, `static` enables multi-NUMA (and single-NUMA) feature, whereas `none` disables it 
+
+Example:
+`--memory-manager-policy=none` or `--memory-manager-policy=static`
+
+#### Pre-Reserved Memory Zone Flag
+
+This will be also followed by a Kubelet Flag for the Memory Manager configuration, which is described above in [The configuration of Memory Manager](#the-configuration-of-memory-manager) section.
+
+`--pre-reserved-memory-zone=[{numa-node=int, memory-type=string, limit=string}][,][...]`
+
+Example:
+   `--pre-reserved-memory-zone= {numa-node=0, memory-type=memory, limit=1Gi}, {numa-node=1, memory-type=memory, limit=1Gi}`
+
+#### Multi-NUMA Groups Flag
+`--multi-numa-groups` - to configure disjoint groups of NUMA nodes, used for preferred multi-NUMA node memory allocation
```

Comment:
I'm still concerned about user experience with this. Specifying groups manually can be hard, especially on a clusters with big number of nodes. I think we should discuss this further.

Would it make sense to consider dynamic configuration? One way of doing this was proposed earlier in the comments: https://github.com/kubernetes/enhancements/pull/1203#discussion_r387660159

Reply:
We dropped this flag from the alpha version, and we generate multi NUMA node hints dynamically. Currently, it returns preferred multi NUMA hint only in case if it no option to satisfy the container requests with the single NUMA node(it means that allocatable of all NUMA nodes less than container requests).

We can re-think the calculation of the multi NUMA preferred hints in the future, but it will be possible only once we will have additional information about the NUMA topology(distances between nodes)

Reply:
Would it make sense to describe dynamic generation of the hints in this document?

As for distances between nodes would [this KEP](https://github.com/kubernetes/enhancements/pull/1718) help with it? If so, should it be referred here?

Reply:
@bart0sh Could you indicate what you exactly understand by "dynamic generation"? I went through the conversation (this one: https://github.com/kubernetes/enhancements/pull/1203#discussion_r387660159 ), but I could only reach something like called "a maximally pessimistic accounting algorithm". Is that what you meant?

Reply:
The idea expressed in https://github.com/kubernetes/enhancements/pull/1203#discussion_r387660159 as far as I understood it is to allocate memory to the minimal possible set of NUMA nodes using 'HW topology tree' approach described in that comment:

- You have the NUMA nodes as the leaves at the bottom in the tree
- above them you have the sockets, each socket having its NUMA nodes as children
- at the root, you have a tree node representing the full machine, with all sockets as children

At the bottom you start with the memory physically attached to the NUMA nodes. As you go up in the tree you always have as available memory the sum of available memory in the children. When you allocate, you assign the container to a tree node where it definitely fits in and you account for the memory allocation at that node as such. In all children of this node allocation you assume the worst case: that the allocation would fully be taken from that single child (because as you yourself noticed/pointed out, you can't know or control where it will take place).

This approach would dynamically define groups of NUMA nodes based on how far NUMA nodes are from each other. It'd also cover both single and cross-numa allocations.

Reply:
Currently, the flow for multi NUMA hints generation is:

1. We check the container requests for memory and hugepages, if it can be satisfied by the one node(it does not mean that it can be satisfied right now because we check allocatable memory) we will set prefer for single NUMA hints only, it means that possible the situation that we will return hints without preferred hints at all, because all preferred hints will be filtered because of lack of memory or hugepages, or because it already used for the multi-numa node allocation for other containers

2. The container requests can not be satisfied with the single NUMA node, in this case, we will check all permutations for NUMA nodes and set preferred to true only for hints that have the minimal number of NUMA nodes that can satisfy the container requests, so if this number 2, we will have hints 0011, 0101, 1001, 0110, 1010, 1100 with preferred field equal to true(again this list can be filtered after it by the memory manager, because of the amount of available memory and hugepages or because one of the nodes under the permutation already used in the single NUMA node allocation for the other container)

3. Because we always return the ordered list, the order that the topology hint will iterate over hints will be 0011, 0101, 1001,...,1100, so it definitely possible that the hint 1001 will be preferred over 1100, but it possible the situation when the hint 1100 will be rejected by the topology manager, and in this case, we still want to use other preferred hints. Currently, the only way to define the preferred hints it is via the `preferred` boolean, if it can be changed to score, we can adapt our algorithm to calculate the score for each hint with the minimal amount of NUMA nodes, but it the subject for the further discussion.